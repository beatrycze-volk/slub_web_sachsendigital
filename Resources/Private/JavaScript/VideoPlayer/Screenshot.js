function renderScreenshot(videoDomElement) {
  // add canvas overlay to DOM
  const domTemplate = document.createElement("template");
  domTemplate.innerHTML = `<div id="screenshot-overlay"><span class="close-screenshot-modal icon-close"></span><canvas></canvas></div>`;
  const domElement = domTemplate.content.firstElementChild;
  document.body.append(domElement);

  const canvas = domElement.querySelector("canvas");
  const closeButton = domElement.querySelector('.close-screenshot-modal');

  // bind close action
  closeButton.addEventListener('click', () => {
    domElement.remove();
  });

  // lets go
  const metadataElement = document.getElementById('metadata');
  const metadataArray = generateMetadataObject(metadataElement);
  drawCanvas(canvas, videoDomElement, metadataArray);
}

/**
 *
 * @param {HTMLCanvasElement | CanvasRenderingContext2D} target Canvas on which the screenshot is drawn
 * @param {HTMLVideoElement} videoDomElement Source video element from which the screenshot is taken
 * @param {any} metadataArray Output of {@link generateMetadataObject}
 */
function drawCanvas(target, videoDomElement, metadataArray) {
  const [targetCanvas, context] =
    target instanceof HTMLCanvasElement
      ? [target, target.getContext('2d')]
      : [target.canvas, target];

  const infoString =
    metadataArray.screenshotFields
      .map(field => metadataArray.metadata[field])
      .filter(value => typeof value === 'string')
      .join(' / ');

  targetCanvas.width = videoDomElement.videoWidth;
  targetCanvas.height = videoDomElement.videoHeight;

  context.drawImage(videoDomElement, 0, 0, targetCanvas.width, targetCanvas.height);

  context.font = '25px Arial';
  context.textAlign = 'end';
  context.fillStyle = "#FFFFFF";
  context.shadowBlur = 5;
  context.shadowColor = "black";
  context.fillText(infoString, targetCanvas.width - 10, targetCanvas.height - 10);

  targetCanvas.style.width = '80%';
  targetCanvas.style.height = 'auto';
}

/**
 *
 * @param {HTMLDataElement} dataDomElement
 */
function generateMetadataObject(dataDomElement) {
  var metadataObject = {
    metadata: {},
    screenshotFields: dataDomElement.getAttribute('data-screenshotfields').split(','),
  };

  for (const child of dataDomElement.children) {
    if (child.value) {
      metadataObject.metadata[child.id] = child.value;
    }
  }

  return metadataObject;
}

module.exports = { drawCanvas, renderScreenshot, generateMetadataObject };
