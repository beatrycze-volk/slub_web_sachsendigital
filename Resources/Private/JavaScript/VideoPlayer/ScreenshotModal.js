// @ts-check

import Environment from './Environment';
import generateTimecodeUrl from './generateTimecodeUrl';
import imageFormats from './image/imageFormats';
import { drawScreenshot } from './Screenshot';
import SimpleModal from './SimpleModal';
import {
  binaryStringToArrayBuffer,
  blobToBinaryString,
  buildTimeString,
  canvasToBlob,
  download,
  e,
  metadataArrayToString,
  sanitizeBasename,
} from './util';

/**
 * @typedef {{
 *  metadata: MetadataArray | null;
 *  showMetadata: boolean;
 *  timecode: number | null;
 *  supportedImageFormats: ImageFormatDesc[];
 *  selectedImageFormat: ImageFormatDesc | null;
 * }} State
 */

/**
 * @extends {SimpleModal<State>}
 */
export default class ScreenshotModal extends SimpleModal {
  /**
   *
   * @param {HTMLElement} parent
   * @param {Environment} env
   */
  constructor(parent, env) {
    const supportedImageFormats = imageFormats.filter(
      format => env.supportsCanvasExport(format.mimeType)
    );

    super(parent, {
      metadata: null,
      showMetadata: true,
      timecode: null,
      supportedImageFormats,
      selectedImageFormat: supportedImageFormats[0] ?? null,
    });

    /** @private */
    this.env = env;
    /** @private @type {HTMLVideoElement | null} */
    this.videoDomElement = null;

    this.$main.classList.add('screenshot-modal');
    this.$title.innerText = env.t('modal.screenshot.title');

    const idShowMetadata = env.mkid();
    const radioGroup = env.mkid();

    this.$body.append(
      e("div", { className: "screenshot-config" }, [
        e("span", { className: "show-metadata" }, [
          e("input", {
            type: "checkbox",
            id: idShowMetadata,
            checked: this.state.showMetadata,
            $change: this.handleChangeShowMetadata.bind(this),
          }),
          e("label", { htmlFor: idShowMetadata }, [
            env.t('modal.screenshot.show-metadata'),
          ]),
        ]),
        " · ",
        e("span", {}, [
          e("label", {}, [env.t('modal.screenshot.file-format')]),
          ":",
          e("span", {}, (
            this.state.supportedImageFormats.flatMap(format => {
              const radioId = env.mkid();

              return [
                e("input", {
                  id: radioId,
                  name: radioGroup,
                  type: 'radio',
                  checked: format.mimeType === this.state.selectedImageFormat?.mimeType,
                  $change: () => {
                    this.setState({
                      selectedImageFormat: format,
                    });
                  },
                }),
                e("label", { htmlFor: radioId }, [` ${format.label}`]),
              ];
            })
          )),
        ]),
        " · ",
        e("a", {
          href: "#",
          $click: this.handleDownloadImage.bind(this),
        }, [env.t('modal.screenshot.download-image')]),
      ]),

      this.$canvas = e("canvas")
    );
  }

  /**
   * Sets video DOM element for upcoming screenshots.
   *
   * @param {HTMLVideoElement} video
   * @returns {this}
   */
  setVideo(video) {
    this.videoDomElement = video;
    return this;
  }

  /**
   * Triggers UI update using new {@link metadata}.
   *
   * @param {MetadataArray} metadata
   * @returns {this}
   */
  setMetadata(metadata) {
    this.setState({ metadata });
    return this;
  }

  /**
   * Triggers UI update using new {@link timecode}.
   *
   * @param {number} timecode
   * @returns {this}
   */
  setTimecode(timecode) {
    this.setState({ timecode });
    return this;
  }

  /**
   * @private
   * @param {Event} e
   */
  handleChangeShowMetadata(e) {
    if (!(e.target instanceof HTMLInputElement)) {
      return;
    }

    this.setState({
      showMetadata: e.target.checked,
    });
  }

  /**
   * @private
   * @param {MouseEvent} e
   */
  async handleDownloadImage(e) {
    e.preventDefault();

    // We could've set `.download-image[href]` in `render()` or in the radio
    // box change listener, but avoid this for performance reasons.

    const { metadata, timecode, selectedImageFormat } = this.state;
    if (metadata === null || timecode === null || selectedImageFormat === null) {
      console.error("one of [metadata, timecode, selectedImageFormat] is null");
      return;
    }

    const image = await this.makeImageBlob(
      this.$canvas, selectedImageFormat, metadata, timecode);
    const filename = this.getFilename(metadata, timecode);

    download(image, filename);
  }

  /**
   *
   * @param {HTMLCanvasElement} canvas
   * @param {ImageFormatDesc} imageFormat
   * @param {MetadataArray} metadata
   * @param {number} timecode
   */
  async makeImageBlob(canvas, imageFormat, metadata, timecode) {
    const imageBlob = await canvasToBlob(canvas, imageFormat.mimeType);
    const imageDataStr = await blobToBinaryString(imageBlob);
    const image = imageFormat.parseBinaryString(imageDataStr);

    if (image) {
      const url = generateTimecodeUrl(timecode, this.env);

      image.addMetadata({
        title: metadata.metadata.title?.[0] ?? "",
        // NOTE: Don't localize (not only relevant to current user)
        comment: `Screenshot taken on Sachsen.Digital.\n\n${url.toString()}`,
      });
      const buffer = binaryStringToArrayBuffer(image.toBinaryString());
      return new Blob([buffer], { type: imageBlob.type });
    } else {
      return imageBlob;
    }
  }

  /**
   *
   * @param {MetadataArray} metadata
   * @param {number} timecode
   */
  getFilename(metadata, timecode) {
    // NOTE: Don't localize (English file name prefix should be alright)
    return sanitizeBasename(
      `Screenshot-${metadata.metadata.title?.[0] ?? "Unnamed"}-T${buildTimeString(timecode, true)}`
    );
  }

  /**
   *
   * @param {MetadataArray | null} metadata
   * @returns {import('./Screenshot').ScreenshotCaption[]}
   */
  getCaptions(metadata) {
    return [
      { v: 'bottom', h: 'left', text: "https://sachsen.digital" },
      { v: 'bottom', h: 'right', text: metadata ? metadataArrayToString(metadata) : "" },
    ];
  }

  /**
   *
   * @param {import('./SimpleModal').BaseState & State} state
   */
  render(state) {
    super.render(state);

    const { show, metadata, showMetadata } = state;

    if (this.videoDomElement === null) {
      // TODO: Error handling
      return;
    }

    if (show && (!this.state.show || showMetadata !== this.state.showMetadata)) {
      const config = {
        captions: showMetadata ? this.getCaptions(metadata) : [],
      };

      drawScreenshot(this.$canvas, this.videoDomElement, config);
    }
  }
}
