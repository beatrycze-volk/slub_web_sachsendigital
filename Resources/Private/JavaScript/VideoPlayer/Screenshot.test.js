/**
 * @jest-environment jsdom
 */

import { drawCanvas } from './Screenshot';
import ScreenshotModal from './ScreenshotModal';

beforeEach(() => {
  // TODO: Reset JSDOM in a more robust way
  document.body.innerHTML = '';
});

function getTestMetadataArray() {
  return {
    metadata: {
      title: "Test Video",
      year: "1912",
      not_a_string: 123,
    },
    screenshotFields: [
      "title",
      "year",
      "not_a_string",
    ],
  };
}

test('can open screenshot overlay', () => {
  const overlay = () => document.querySelector('.screenshot-modal');

  window.VIDEO = {
    metadata: getTestMetadataArray(),
  };

  // not opened yet
  expect(overlay()).toBeNull();

  // opened; exact tags are in snapshot
  const modal = new ScreenshotModal(document.body, new VideoMock(1920, 1080));
  modal.open();
  expect(overlay()).toMatchSnapshot();
});

class VideoMock extends HTMLVideoElement {
  constructor(width, height) {
    super();

    this._mockWidth = width;
    this._mockHeight = height;
  }

  get videoWidth() {
    return this._mockWidth;
  }

  get videoHeight() {
    return this._mockHeight;
  }
}

customElements.define('video-mock', VideoMock, { extends: 'video' });

test('can draw to canvas', () => {
  const metadata = getTestMetadataArray();

  const snapshotWithSize = (videoWidth, videoHeight) => {
    const video = new VideoMock(videoWidth, videoHeight);

    const canvas = document.createElement("canvas");
    const context = canvas.getContext('2d');

    drawCanvas(context, video, metadata);

    const events = context.__getEvents();
    expect(events).toMatchSnapshot();
  };

  snapshotWithSize(1920, 1080);
  snapshotWithSize(960, 540);
});
