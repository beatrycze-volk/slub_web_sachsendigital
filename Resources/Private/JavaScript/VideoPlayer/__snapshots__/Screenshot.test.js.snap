// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can draw to canvas 1`] = `
Array [
  Object {
    "props": Object {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 0,
      "dy": 0,
      "img": VideoMock {
        "_mockHeight": 1080,
        "_mockWidth": 1920,
      },
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "value": "25px Arial",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "end",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": 5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "shadowBlur",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "shadowColor",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "Test Video / 1912",
      "x": 1910,
      "y": 1070,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;

exports[`can draw to canvas 2`] = `
Array [
  Object {
    "props": Object {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 0,
      "dy": 0,
      "img": VideoMock {
        "_mockHeight": 540,
        "_mockWidth": 960,
      },
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "value": "12px Arial",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "end",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
  Object {
    "props": Object {
      "value": 5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "shadowBlur",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "shadowColor",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "Test Video / 1912",
      "x": 955,
      "y": 535,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;

exports[`can open screenshot overlay 1`] = `
<div
  class="sxnd-modal screenshot-modal"
>
  <div
    class="headline-container"
  >
    <h3 />
    <span
      class="modal-close icon-close"
    />
  </div>
  <div
    class="body-container"
  >
    <div
      class="screenshot-config"
    >
      
        
      <input
        class="show-metadata"
        id="cb-show-metadata"
        type="checkbox"
      />
      <!--
        -->
      <label
        for="cb-show-metadata"
      >
         Metadaten einblenden
      </label>
      

        Â·

        
      <a
        class="download-image"
        href="#"
      >
        Bild herunterladen
      </a>
      
      
    </div>
    <canvas
      height="1080"
      width="1920"
    />
  </div>
</div>
`;
